benchmarks/README.md

```diff
@@ -0,0 +1,104 @@
+export ENVOY_DOCKER_IMAGE_TO_TEST="envoyproxy/envoy-dev:74290ef76a76fbbf50f072dc33438791f93f68c7"
+# Envoy is called 'Envoy' in the Envoy docker image.
+export ENVOY_PATH="envoy"
+
+# run all tests starting with test_http_h1_small in benchmarks/
+bazel-bin/benchmarks/benchmarks --log-cli-level=info -vvvv -k test_http_h1_small benchmarks/
+```
+
+## Example: running with binaries
+
+This will build the Nighthawk binaries from the c++ code, and use those to
+execute the benchmarks. Environment variable `ENVOY_PATH` can be used to
+specify a custom Envoy binary to use to inject as a proxy between the test
+client and server. If not set, the benchmark suite will fall back to configuring
+Nighthawk's test server for that. Note that the build can be a lengthy process.
+
+```bash
+git clone https://github.com/oschaaf/nighthawk.git benchmark-test
+cd benchmark-test
+bazel test --test_summary=detailed --test_output=all --test_arg=--log-cli-level=info --test_env=ENVOY_IP_TEST_VERSIONS=v4only --test_env=HEAPPROFILE= --test_env=HEAPCHECK= --cache_test_results=no --compilation_mode=opt --cxxopt=-g --cxxopt=-ggdb3 //benchmarks:*
+```
+
+## Example: fully dockerized flow
+
+The framework can be run via docker and used that way to execute
+python benchmarks scripts not sourced from this repository, but
+elsewhere. An example:
+
+```bash
+# This script runs the dockerized benchmarking framework, which in
+# turn will pull Nighthawk and Envoy in via docker. 
+
+set -eo pipefail
+set +x
+set -u
+
+# The benchmark logs and artifacts will be dropped here
+OUTDIR="/home/oschaaf/code/envoy-perf-vscode/nighthawk/benchmarks/tmp/"
+# Used to map the test that we want to see executed into the docker container
+# Note: the contents could be fetched via http, for example.
+TEST_DIR="/home/oschaaf/code/envoy-perf-vscode/nighthawk/benchmarks/test/"
+
+# Rebuild the docker in case something changed.
+./docker_build.sh && 
+docker run -it --rm \
+  -v "/var/run/docker.sock:/var/run/docker.sock:rw" \
+  -v "${OUTDIR}:${OUTDIR}:rw" \
+  -v "${TEST_DIR}:/usr/local/bin/benchmarks/benchmarks.runfiles/nighthawk/benchmarks/external_tests/" \
+  --network=host \
+  --env NH_DOCKER_IMAGE="envoyproxy/nighthawk-dev:latest" \
+  --env ENVOY_DOCKER_IMAGE_TO_TEST="envoyproxy/envoy-dev:f61b096f6a2dd3a9c74b9a9369a6ea398dbe1f0f" \
+  --env TMPDIR="${OUTDIR}" \
+  oschaaf/benchmark-dev:latest ./benchmarks --log-cli-level=info -vvvv 
+```
+
+# TODOs
+
+- Copy out the artifacts and push those to a gcp bucket. Current status:
+   - cpu profiles are dumped to tmp per test (named according to the test). ideally we'd
+     also dump flamegraph svg's
+- A UI -- though we may be able to get by with just a uri structure conventioned around the envoy
+  sha. e.g. http://perf-ci-host/gcpsync/[envoy-sha]-[timestamp]/ to link CI, and directory index the 
+  artifacts.
+
+# FUTURE
+
+- profiling / flamegraphing via perf/bcc tools; include the proxy-wasm flamegraphing research r&d
```

Comment:
I think the most useful things to add are (in my $0.02 stack rank):
1. The ability to repeat the runs multiple times and obtain stats, e.g. how much variance there is, mean, etc.
2. The ability to obtain flamegraphs.
3. The ability to do A/B testing, similar to https://github.com/envoyproxy/envoy-perf/blob/master/siege/siege.py#L3.

But, I'm sure there are other folks who have other priorities here ;-)

Reply:
Also moved these into the future list in README.md